presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinGraphNode.java

```diff
@@ -0,0 +1,151 @@
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static com.facebook.presto.sql.ExpressionUtils.and;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class represents a set of inner joins that can be executed in any order.
+ * The nodes of the join graph are the sources to be joined. The edges are
+ * filters. If there is no filter between a pair
+ * of sources, then a possible cross join is implied
+ */
+public class JoinGraphNode
+        extends PlanNode
+{
+    private final List<PlanNode> sources;
+    private final Expression filter;
+    private final List<Symbol> outputSymbols;
+
+    public JoinGraphNode(PlanNodeId id, List<PlanNode> sources, Expression filter, List<Symbol> outputSymbols)
+    {
+        super(id);
+
+        requireNonNull(sources, "sources is null");
+        requireNonNull(filter, "filters is null");
+
+        this.sources = ImmutableList.copyOf(sources);
+        this.filter = filter;
+        this.outputSymbols = ImmutableList.copyOf(outputSymbols);
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    @Override
+    public <R, C> R accept(PlanVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitJoinGraph(this, context);
+    }
+
+    @Override
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    @Override
+    public List<Symbol> getOutputSymbols()
+    {
+        return outputSymbols;
+    }
+
+    @Override
+    public PlanNode replaceChildren(List<PlanNode> newChildren)
+    {
+        return new JoinGraphNode(getId(), newChildren, filter, outputSymbols);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(this.getOutputSymbols(), this.getFilter(), this.getSources().stream().map(PlanNode::getId).collect(toImmutableList()));
+    }
+
+    @Override
+    public boolean equals(Object other)
+    {
+        if (!(other instanceof JoinGraphNode)) {
+            return false;
+        }
+
+        JoinGraphNode otherJoinGraph = (JoinGraphNode) other;
+        return this.getOutputSymbols().equals(otherJoinGraph.getOutputSymbols())
+                && this.getFilter().equals(otherJoinGraph.getFilter())
+                && this.getSources().stream().map(PlanNode::getId).collect(toImmutableList()).equals(otherJoinGraph.getSources().stream().map(PlanNode::getId));
```

Comment:
I didn't notice this will be always original sources IDs. I know there are JoinGraphNodes that have other JoinGraphNodes as sources (so they are not "original" and have generated IDs), but maybe they can never reach memo.